File: server/app/data/generate_org_tree.py
import random
from dataclasses import dataclass
from datetime import datetime, timedelta, timezone

from app.models import OrgNode

DIVISION_NAMES = [
    "Дивизион продаж",
    "Дивизион разработки",
    "Дивизион маркетинга",
    "Дивизион операций",
    "Дивизион финансов",
]

DEPARTMENT_NAMES = [
    "Отдел клиентского сервиса",
    "Отдел аналитики",
    "Отдел разработки продукта",
    "Отдел контроля качества",
    "Отдел закупок",
    "Отдел кадров",
    "Отдел логистики",
    "Отдел бухгалтерии",
]

TEAM_NAMES = [
    "Команда фронтенда",
    "Команда бэкенда",
    "Команда инфраструктуры",
    "Команда дизайна",
    "Команда поддержки",
    "Команда исследований",
    "Команда продаж B2B",
    "Команда продаж B2C",
]


def _pick_name(pool: list[str], index: int) -> str:
    base = pool[index % len(pool)]
    cycle = index // len(pool)
    return base if cycle == 0 else f"{base} {cycle + 1}"


def _random_updated_at() -> str:
    days_ago = random.randint(0, 30)
    moment = datetime.now(timezone.utc) - timedelta(days=days_ago)
    return moment.isoformat().replace("+00:00", "Z")


@dataclass
class GenerateOptions:
    divisions_count: int = 5
    departments_per_division: tuple[int, int] = (3, 4)
    teams_per_department: tuple[int, int] = (2, 4)


def generate_org_tree(options: GenerateOptions | None = None) -> list[OrgNode]:
    options = options or GenerateOptions()

    nodes: list[OrgNode] = []
    next_id = 1

    def make_node(parent_id: str | None, name: str) -> OrgNode:
        nonlocal next_id
        node = OrgNode(
            id=str(next_id),
            name=name,
            parentId=parent_id,
            headcount=random.randint(2, 40),
            budget=random.randint(500_000, 20_000_000),
            performance=random.randint(0, 100),
            updatedAt=_random_updated_at(),
        )
        next_id += 1
        return node

    for d in range(options.divisions_count):
        division = make_node(None, _pick_name(DIVISION_NAMES, d))
        nodes.append(division)

        departments_count = random.randint(*options.departments_per_division)

        for dep in range(departments_count):
            department = make_node(division.id, _pick_name(DEPARTMENT_NAMES, dep))
            nodes.append(department)

            teams_count = random.randint(*options.teams_per_department)

            for t in range(teams_count):
                nodes.append(make_node(department.id, _pick_name(TEAM_NAMES, t)))

    return nodes


@dataclass
class IntegrityCheckResult:
    is_valid: bool
    errors: list[str]


def check_org_tree_integrity(nodes: list[OrgNode]) -> IntegrityCheckResult:
    errors: list[str] = []
    ids = {node.id for node in nodes}

    for node in nodes:
        if node.parent_id is not None and node.parent_id not in ids:
            errors.append(
                f"Узел {node.id} ссылается на несуществующего родителя {node.parent_id}"
            )

    return IntegrityCheckResult(is_valid=len(errors) == 0, errors=errors)

File: server/src/data/generateOrgTree.ts
import type { OrgNode } from "../types.js";

const DIVISION_NAMES = [
  "Дивизион продаж",
  "Дивизион разработки",
  "Дивизион маркетинга",
  "Дивизион операций",
  "Дивизион финансов",
];

const DEPARTMENT_NAMES = [
  "Отдел клиентского сервиса",
  "Отдел аналитики",
  "Отдел разработки продукта",
  "Отдел контроля качества",
  "Отдел закупок",
  "Отдел кадров",
  "Отдел логистики",
  "Отдел бухгалтерии",
];

const TEAM_NAMES = [
  "Команда фронтенда",
  "Команда бэкенда",
  "Команда инфраструктуры",
  "Команда дизайна",
  "Команда поддержки",
  "Команда исследований",
  "Команда продаж B2B",
  "Команда продаж B2C",
];

function randomInt(min: number, max: number): number {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

function pickName(pool: string[], index: number): string {
  const base = pool[index % pool.length];
  const cycle = Math.floor(index / pool.length);
  return cycle === 0 ? base : `${base} ${cycle + 1}`;
}

function randomUpdatedAt(): string {
  const now = Date.now();
  const daysAgo = randomInt(0, 30);
  return new Date(now - daysAgo * 24 * 60 * 60 * 1000).toISOString();
}

interface GenerateOptions {
  divisionsCount?: number;
  departmentsPerDivision?: [number, number];
  teamsPerDepartment?: [number, number];
}

export function generateOrgTree(options: GenerateOptions = {}): OrgNode[] {
  const {
    divisionsCount = 5,
    departmentsPerDivision = [3, 4],
    teamsPerDepartment = [2, 4],
  } = options;

  const nodes: OrgNode[] = [];
  let nextId = 1;

  const makeNode = (parentId: string | null, name: string): OrgNode => {
    const id = String(nextId++);
    return {
      id,
      name,
      parentId,
      headcount: randomInt(2, 40),
      budget: randomInt(500_000, 20_000_000),
      performance: randomInt(0, 100),
      updatedAt: randomUpdatedAt(),
    };
  };

  for (let d = 0; d < divisionsCount; d++) {
    const division = makeNode(null, pickName(DIVISION_NAMES, d));
    nodes.push(division);

    const departmentsCount = randomInt(
      departmentsPerDivision[0],
      departmentsPerDivision[1],
    );

    for (let dep = 0; dep < departmentsCount; dep++) {
      const department = makeNode(
        division.id,
        pickName(DEPARTMENT_NAMES, dep),
      );
      nodes.push(department);

      const teamsCount = randomInt(
        teamsPerDepartment[0],
        teamsPerDepartment[1],
      );

      for (let t = 0; t < teamsCount; t++) {
        nodes.push(makeNode(department.id, pickName(TEAM_NAMES, t)));
      }
    }
  }

  return nodes;
}

export interface IntegrityCheckResult {
  isValid: boolean;
  errors: string[];
}

export function checkOrgTreeIntegrity(nodes: OrgNode[]): IntegrityCheckResult {
  const errors: string[] = [];
  const ids = new Set(nodes.map((n) => n.id));

  for (const node of nodes) {
    if (node.parentId !== null && !ids.has(node.parentId)) {
      errors.push(
        `Узел ${node.id} ссылается на несуществующего родителя ${node.parentId}`,
      );
    }
  }

  return { isValid: errors.length === 0, errors };
}
